File: Milestone-3/AI-voice-mental-Health-Assistant/backend/safety_monitor.py
"""
Crisis Detection and Safety Layer - Backend Module
Real-time monitoring for crisis indicators and emergency response
"""

import json
import os
import re
from typing import Dict, List, Optional, Tuple
from enum import Enum
from dataclasses import dataclass
from datetime import datetime


class CrisisLevel(Enum):
    """Crisis severity levels"""
    NONE = "NONE"
    WARNING = "WARNING"
    CRITICAL = "CRITICAL"


@dataclass
class CrisisResponse:
    """Crisis detection response data"""
    level: CrisisLevel
    triggered_words: List[str]
    message: str
    timestamp: datetime
    requires_override: bool


class SafetyMonitor:
    """Backend crisis detection and safety monitoring system"""
    
    def __init__(self, config_path: str = "botConfig.json"):
        """
        Initialize safety monitor with configuration
        """
        self.config = self._load_config(config_path)
        
        # Critical trigger words from config
        self.critical_triggers = self.config.get("crisis", {}).get("keywords", [])
        self.crisis_response_message = self.config.get("crisis", {}).get("response", "")
        
        # PHQ-9 Question 9 specific patterns
        # Keeping these regexes as they are robust for catching variants
        self.phq9_q9_positive_patterns = [
            r'\b(yes|sometimes|often|nearly every day|several days)\b',
            r'\b(more than half|have thought|been thinking)\b',
            r'\b(crossed my mind|considered|considering|thinking about)\b',
            r'\b(thoughts? of|idea of|wanting to)\b.*\b(die|death|harm)\b'
        ]
        
        self.phq9_q9_negative_patterns = [
            r'\b(no|never|not at all|haven\'t|don\'t)\b',
            r'\b(absolutely not|definitely not|of course not)\b'
        ]
    
    def _load_config(self, config_path: str) -> Dict:
        """Load bot configuration from JSON file"""
        try:
            # Try absolute path or relative to backend
            paths_to_try = [
                config_path,
                os.path.join("backend", config_path),
                os.path.join(os.path.dirname(__file__), config_path)
            ]
            
            for path in paths_to_try:
                if os.path.exists(path):
                    with open(path, 'r', encoding='utf-8') as f:
                        return json.load(f)
            
            print(f"Warning: Could not find {config_path}. Using empty config.")
            return {}
        except Exception as e:
            print(f"Error loading config: {e}")
            return {}

    def check_for_crisis(self, text: str) -> CrisisResponse:
        """
        Analyze text for crisis indicators
        """
        if not text or not isinstance(text, str):
            return CrisisResponse(
                level=CrisisLevel.NONE,
                triggered_words=[],
                message="",
                timestamp=datetime.now(),
                requires_override=False
            )
        
        normalized_text = text.lower().strip()
        triggered_words = []
        
        # Check for critical triggers
        for trigger in self.critical_triggers:
            if trigger.lower() in normalized_text:
                triggered_words.append(trigger)
        
        if triggered_words:
            return CrisisResponse(
                level=CrisisLevel.CRITICAL,
                triggered_words=triggered_words,
                message=self._get_crisis_message(),
                timestamp=datetime.now(),
                requires_override=True
            )
        
        return CrisisResponse(
            level=CrisisLevel.NONE,
            triggered_words=[],
            message="",
            timestamp=datetime.now(),
            requires_override=False
        )
    
    def check_phq9_question9(self, response: str) -> bool:
        """
        Check PHQ-9 Question 9 (suicidal ideation) for positive response
        """
        if not response or not isinstance(response, str):
            return False
        
        normalized_response = response.lower().strip()
        
        # Check for clear negative responses first
        for pattern in self.phq9_q9_negative_patterns:
            if re.search(pattern, normalized_response, re.IGNORECASE):
                if len(normalized_response) < 20:
                    return False
        
        # Check for positive indicators
        for pattern in self.phq9_q9_positive_patterns:
            if re.search(pattern, normalized_response, re.IGNORECASE):
                return True
        
        return False
    
    def _get_crisis_message(self) -> str:
        """Get the crisis response message"""
        return self.crisis_response_message


class CrisisOverrideHandler:
    """Handles crisis response overrides and emergency protocols"""
    
    def __init__(self, safety_monitor: SafetyMonitor):
        self.safety_monitor = safety_monitor
        self.crisis_detected = False
        self.crisis_timestamp = None
    
    def process_user_input(self, text: str, is_phq9_q9: bool = False) -> Tuple[bool, Optional[str]]:
        """
        Process user input and determine if crisis override is needed
        """
        # Check for general crisis indicators
        crisis_response = self.safety_monitor.check_for_crisis(text)
        
        # Check PHQ-9 Question 9 specifically
        if is_phq9_q9 and self.safety_monitor.check_phq9_question9(text):
            self.crisis_detected = True
            self.crisis_timestamp = datetime.now()
            return True, self.safety_monitor._get_crisis_message()
        
        # Handle critical crisis detection
        if crisis_response.level == CrisisLevel.CRITICAL:
            self.crisis_detected = True
            self.crisis_timestamp = datetime.now()
            return True, crisis_response.message
        
        return False, None
    
    def is_crisis_active(self) -> bool:
        """Check if crisis state is currently active"""
        return self.crisis_detected
    
    def reset_crisis_state(self) -> None:
        """Reset crisis detection state"""
        self.crisis_detected = False
        self.crisis_timestamp = None

File: Milestone 4/Infosys_-Project_FitPlan-AI-Personalized-Fitness-Plan-Generato-main/Infosys_-Project_FitPlan-AI-Personalized-Fitness-Plan-Generato-main/Frontend/public/ai-assistant/js/safety-monitor.js
/**
 * Safety Monitor - Crisis Detection System
 * Real-time monitoring for crisis indicators and trigger words
 */

class SafetyMonitor {
    constructor(customTriggerWords = []) {
        // Crisis trigger words (configurable)
        this.triggerWords = [
            'suicide', 'suicidal', 'kill myself', 'end my life',
            'self-harm', 'hurt myself', 'cut myself', 'harm myself',
            'want to die', 'better off dead', 'no point living',
            'hurt others', 'kill someone', 'harm others', 'end it all',
            'take my own life', 'not worth living', 'kill them'
        ];
        
        // Warning indicators (less severe)
        this.warningWords = [
            'hopeless', 'worthless', 'trapped', 'burden',
            'desperate', 'overwhelmed', 'can\'t cope', 'give up',
            'no way out', 'pointless', 'useless', 'hate myself'
        ];
        
        // Add custom trigger words if provided
        if (Array.isArray(customTriggerWords)) {
            this.triggerWords.push(...customTriggerWords);
        }
        
        this.crisisDetected = false;
        this.crisisTimestamp = null;
        this.overrideCallbacks = [];
    }
    
    /**
     * Check text for crisis indicators
     * @param {string} text - User input text to analyze
     * @returns {string} Crisis level: 'NONE', 'WARNING', 'CRITICAL'
     */
    checkForCrisis(text) {
        if (!text || typeof text !== 'string') {
            return 'NONE';
        }
        
        const normalizedText = text.toLowerCase().trim();
        
        // Check for critical trigger words
        const hasCriticalTrigger = this.triggerWords.some(trigger => 
            normalizedText.includes(trigger.toLowerCase())
        );
        
        if (hasCriticalTrigger) {
            this.crisisDetected = true;
            console.warn('CRISIS DETECTED: Critical trigger words found');
            return 'CRITICAL';
        }
        
        // Check for warning indicators
        const hasWarningIndicator = this.warningWords.some(warning => 
            normalizedText.includes(warning.toLowerCase())
        );
        
        if (hasWarningIndicator) {
            console.warn('WARNING: Concerning language detected');
            return 'WARNING';
        }
        
        return 'NONE';
    }
    
    /**
     * Check if PHQ-9 Question 9 (suicidal ideation) has positive response
     * @param {string} response - User response to analyze
     * @returns {boolean} True if positive response detected
     */
    checkPHQ9Question9(response) {
        if (!response || typeof response !== 'string') {
            return false;
        }
        
        const normalizedResponse = response.toLowerCase().trim();
        
        // Positive indicators for suicidal ideation
        const positiveIndicators = [
            'yes', 'sometimes', 'often', 'nearly every day',
            'several days', 'more than half', 'have thought',
            'been thinking', 'crossed my mind', 'considered'
        ];
        
        // Negative indicators (clear denial)
        const negativeIndicators = [
            'no', 'never', 'not at all', 'haven\'t', 'don\'t',
            'absolutely not', 'definitely not'
        ];
        
        // Check for clear negative responses first
        const hasNegativeIndicator = negativeIndicators.some(negative => 
            normalizedResponse.includes(negative)
        );
        
        if (hasNegativeIndicator && normalizedResponse.length < 20) {
            return false; // Short, clear negative response
        }
        
        // Check for positive indicators
        const hasPositiveIndicator = positiveIndicators.some(positive => 
            normalizedResponse.includes(positive)
        );
        
        if (hasPositiveIndicator) {
            this.crisisDetected = true;
            console.warn('PHQ-9 Q9 POSITIVE: Suicidal ideation detected');
            return true;
        }
        
        return false;
    }
    
    /**
     * Get crisis response message
     * @returns {string} Appropriate crisis response message
     */
    getCrisisResponseMessage() {
        return `I'm very concerned about what you've shared with me. Your safety is the most important thing right now. 

Please reach out for immediate support:
• Crisis Helpline: 988 (available 24/7)
• Emergency Services: 911
• Crisis Text Line: Text HOME to 741741

You don't have to go through this alone. There are people who want to help you, and things can get better. Please reach out to one of these resources right away.`;
    }
    
    /**
     * Get warning response message for concerning but non-critical content
     * @returns {string} Appropriate warning response message
     */
    getWarningResponseMessage() {
        return `I hear that you're going through a difficult time. While I can help with this screening, please remember that professional support is available if you need someone to talk to. 

If you're feeling overwhelmed, consider reaching out to:
• Crisis Helpline: 988
• Your healthcare provider
• A trusted friend or family member

Would you like to continue with the screening, or would you prefer information about mental health resources?`;
    }
    
    /**
     * Register callback for crisis override mechanism
     * @param {Function} callback - Function to call when crisis override is needed
     */
    registerOverrideCallback(callback) {
        if (typeof callback === 'function') {
            this.overrideCallbacks.push(callback);
        }
    }
    
    /**
     * Trigger crisis response override
     * Calls all registered override callbacks and terminates normal flow
     */
    triggerCrisisOverride() {
        this.crisisDetected = true;
        this.crisisTimestamp = new Date();
        
        console.error('CRISIS OVERRIDE TRIGGERED - Terminating normal conversation flow');
        
        // Call all registered override callbacks
        this.overrideCallbacks.forEach(callback => {
            try {
                callback(this.getCrisisResponseMessage());
            } catch (error) {
                console.error('Error in crisis override callback:', error);
            }
        });
    }
    
    /**
     * Process user input and determine if override is needed
     * @param {string} text - User input text
     * @param {boolean} isPhq9Q9 - Whether this is PHQ-9 Question 9
     * @returns {Object} Processing result with override status
     */
    processUserInput(text, isPhq9Q9 = false) {
        const crisisLevel = this.checkForCrisis(text);
        
        // Check PHQ-9 Question 9 specifically
        if (isPhq9Q9 && this.checkPHQ9Question9(text)) {
            this.triggerCrisisOverride();
            return {
                shouldOverride: true,
                level: 'CRITICAL',
                message: this.getCrisisResponseMessage(),
                triggeredBy: 'PHQ9_Q9'
            };
        }
        
        // Handle critical crisis detection
        if (crisisLevel === 'CRITICAL') {
            this.triggerCrisisOverride();
            return {
                shouldOverride: true,
                level: 'CRITICAL',
                message: this.getCrisisResponseMessage(),
                triggeredBy: 'TRIGGER_WORDS'
            };
        }
        
        // Handle warning level (no override, but return warning)
        if (crisisLevel === 'WARNING') {
            return {
                shouldOverride: false,
                level: 'WARNING',
                message: this.getWarningResponseMessage(),
                triggeredBy: 'WARNING_WORDS'
            };
        }
        
        return {
            shouldOverride: false,
            level: 'NONE',
            message: null,
            triggeredBy: null
        };
    }
    
    /**
     * Reset crisis detection state
     */
    reset() {
        this.crisisDetected = false;
        this.crisisTimestamp = null;
    }
    
    /**
     * Check if crisis state is currently active
     * @returns {boolean} True if crisis was detected and is still active
     */
    isCrisisActive() {
        return this.crisisDetected;
    }
    
    /**
     * Get crisis timestamp
     * @returns {Date|null} Timestamp when crisis was detected
     */
    getCrisisTimestamp() {
        return this.crisisTimestamp;
    }
    
    /**
     * Add custom trigger words (for configuration)
     * @param {string[]} words - Array of trigger words to add
     */
    addTriggerWords(words) {
        if (Array.isArray(words)) {
            this.triggerWords.push(...words);
        }
    }
    
    /**
     * Add custom warning words (for configuration)
     * @param {string[]} words - Array of warning words to add
     */
    addWarningWords(words) {
        if (Array.isArray(words)) {
            this.warningWords.push(...words);
        }
    }
}